### ISSUE TITLE
Task.send((String mailbox, double timeout)) maybe has an error

### ISSUE THREAD

**Issue by kenenbek** [mquinson/simgrid#72]
I study the examples of simgrid-java. [In master-worker](https://github.com/mquinson/simgrid/blob/master/examples/java/app/masterworker/Master.java) there is a Task.send(mailbox) method, so it works fine. But when I want that master distributes tasks with some delays the using of method send(String mailbox, double timeout) gives me an error. I modified the code from Master.java as such:
```
    for (int i = 0; i < workersCount; i++) {
      Task task = new Task("finalize", 0, 0);
      task.send("worker_"+(i%workersCount), 1.1);
    }
```

```
[Jackson:masterslave.Forwarder:(2) 0.851467] [jmsg/INFO] Sending "Task_0" to "slave_7"
org.simgrid.msg.TimeoutException
	at org.simgrid.msg.Task.receive(Native Method)
	at org.simgrid.msg.Task.receive(Task.java:266)
	at masterslave.Slave.main(Slave.java:32)
	at org.simgrid.msg.Process.run(Process.java:341)
[Stephen:masterslave.Slave:(9) 1.951467] [jmsg/INFO] Unexpected behavior. Stopping now

 A fatal error has been `detected` by the Java Runtime Environment:

SIGSEGV (0xb) at pc=0x00007f3598471f98, pid=10660, tid=139868434626304
JRE version: OpenJDK Runtime Environment (7.0_95) (build 1.7.0_95-b00)
 Java VM: OpenJDK 64-Bit Server VM (24.95-b01 mixed mode linux-amd64 compressed oops)
Derivative: IcedTea 2.6.4
Distribution: Ubuntu 14.04.4 LTS, package 7u95-2.6.4-0ubuntu0.14.04.1ppa1
Problematic frame:
C  [libsimgrid.so+0x8bf98]
Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
```

**Comment by frs69wq** [comment]
Hi,
Copy/Pasting part of the output log is not enough to get help. You need to provide a minimal working example that reproduces what you think is a bug. 
In your case, as you modified one of the SimGrid examples, the least you can do is to give the diff of your modifications.
My guess is that for this particular communication between Jackson and Stephen, the timeout expires and exception is raised but not properly catched.

**Comment by kenenbek** [comment]
I updated.

**Comment by frs69wq** [comment]
The send starts at 0.851467, the timeout expires after 1.1 seconds and 
the exception is then raised at 1.951467, when the error occurs. This 
just means that the exception is not catched in your code and then makes 
the execution stops.

This is not a bug, but a problem in your modified code. If you want your 
code to react to the expiration of a timeout, e.g., resend the task with 
a larger timeout or do anything else, you have to add a try/catch block 
and add some code to do something.



Le 02/04/2016 15:16, kenenbek a écrit :
>
> I study the examples of simgrid-java. In master-worker 
> <https://github.com/mquinson/simgrid/blob/master/examples/java/app/masterworker/Master.java> 
> there is a Task.send(mailbox) method, so it works fine. But when I 
> want that master distributes tasks with some delays the using of 
> method send(String mailbox, double timeout) gives me an error.
> `/usr/lib/jvm/java-7-openjdk-amd64/bin/java -Didea.launcher.port=7532 
> -Didea.launcher.bin.path=/home/ken/Downloads/idea-IU-145.597.3/bin 
> -Dfile.encoding=UTF-8 -classpath 
> /usr/lib/jvm/java-7-openjdk-amd64/jre/lib/charsets.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/compilefontconfig.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/ext/dnsns.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/ext/icedtea-sound.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/ext/java-atk-wrapper.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/ext/localedata.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/ext/sunjce_provider.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/ext/sunpkcs11.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/ext/zipfs.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/javazic.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/jce.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/jsse.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/management-agent.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/resources.j 
> ar:/usr/ 
> lib/jvm/java-7-openjdk-amd64/jre/lib/rhino.jar:/usr/lib/jvm/java-7-openjdk-amd64/jre/lib/rt.jar:/home/ken/Downloads/SimGrid-3.12/examples/java/async/production/java:/home/ken/Downloads/simgrid.jar:/home/ken/Downloads/idea-IU-145.597.3/lib/idea_rt.jar 
> com.intellij.rt.execution.application.AppMain masterslave.Masterslave 
> /home/ken/Downloads/SimGrid-3.12/examples/platforms/platform.xml 
> /home/ken/Downloads/SimGrid-3.12/examples/java/masterslave/masterslaveDeployment.xml
> [0.000000] [jmsg/INFO] Using regular java threads. Coroutines could 
> speed your simulation up.
> [Jacquelin:masterslave.Master:(1) 0.000000] [jmsg/INFO] Hello! Got 7 
> slaves and 5 tasks to process
> [Jackson:masterslave.Forwarder:(2) 0.000000] [jmsg/INFO] Receiving on 
> 'slave_0'
> [Casavant:masterslave.Forwarder:(3) 0.000000] [jmsg/INFO] Receiving on 
> 'slave_1'
> [Jackson:masterslave.Forwarder:(2) 0.851467] [jmsg/INFO] Sending 
> "Task_0" to "slave_7"
> org.simgrid.msg.TimeoutException
> at org.simgrid.msg.Task.receive(Native Method)
> at org.simgrid.msg.Task.receive(Task.java:266)
> at masterslave.Slave.main(Slave.java:32)
> at org.simgrid.msg.Process.run(Process.java:341)
> [Stephen:masterslave.Slave:(9) 1.951467] [jmsg/INFO] Unexpected 
> behavior. Stopping now
> #
>
>
>   A fatal error has been detected by the Java Runtime Environment:
>
> #
>
>
>   SIGSEGV (0xb) at pc=0x00007f3598471f98, pid=10660, tid=139868434626304
>
> #
>
>
>   JRE version: OpenJDK Runtime Environment (7.0_95) (build 1.7.0_95-b00)
>
>
>   Java VM: OpenJDK 64-Bit Server VM (24.95-b01 mixed mode linux-amd64
>   compressed oops)
>
>
>   Derivative: IcedTea 2.6.4
>
>
>   Distribution: Ubuntu 14.04.4 LTS, package
>   7u95-2.6.4-0ubuntu0.14.04.1ppa1
>
>
>   Problematic frame:
>
>
>   C [libsimgrid.so+0x8bf98]
>
> #
>
>
>   Failed to write core dump. Core dumps have been disabled. To enable
>   core dumping, try "ulimit -c unlimited" before starting Java again`
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/mquinson/simgrid/issues/72>
>

-- 
One should, for example, be able to see that things are hopeless
and yet be determined to make them otherwise.
                                           -- F. Scott Fitzgerald

**Comment by kenenbek** [comment]
Could you advise me a construction I should use for situation when master has to send tasks to worker every  100 sec?

**Comment by kenenbek** [comment]
For delays I simply can use waitFor(double timeout).

**Comment by mquinson** [comment]
In this situation, you should use non-blocking communication. Check the examples/msg/async-* examples.
As a side note, could you repost this specific question on stackoverflow and point us to it?

**Comment by mquinson** [comment]
This is duplicated here: http://stackoverflow.com/questions/36485083/simgrid-java-sending-instructions-from-master-to-worker 
thus closing. 
Thanks @kenenbek for moving this discussion to the right location.
